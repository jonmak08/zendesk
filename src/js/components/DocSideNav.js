import preact from 'preact';

import * as apiHelpers from '../helpers/api-helpers.js';
import LoadingIndicator from './LoadingIndicator.js';

class ArticlesList extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);

		this.state = {
			expanded: false,
			items: [],
			loading: false
		};
	}

	getArticles() {
		const {locale, id} = this.props;

		this.setState({loading: true});

		apiHelpers
			.getArticlesBySectionId(locale, id)
			.then(({data}) => {
				this.setState({
					items: data,
					loading: false
				});
			})
			.catch(err => this.setState({loading: false}));
	}

	handleClick() {
		this.setState({expanded: !this.state.expanded});

		if (!this.state.items.length) {
			this.getArticles();
		}
	}

	render({name}, {expanded, items, loading}) {
		return (
			<div>
				<a
					class="sidenav-item"
					href="javascript:;"
					onClick={this.handleClick}
				>
					{name}
				</a>

				{expanded &&
					!loading && (
						<ul class="nav nav-nested">
							{items.articles.map(item => (
								<li class="nav-item" key={item.id}>
									<a
										class="sidenav-item"
										href={item.html_url}
									>
										{item.name}
									</a>
								</li>
							))}
						</ul>
					)}

				{loading && <LoadingIndicator />}
			</div>
		);
	}
}

export default class DocSideNav extends preact.Component {
	constructor(props) {
		super(props);

		this.state = {
			items: [],
			loading: true
		};
	}

	componentDidMount() {
		const {locale} = this.props;

		apiHelpers
			.getSectionsCategories(locale)
			.then(({data}) => {
				let productDocCategory = data.categories.find(
					item => item.name === 'Analytics Cloud Documentation'
				);

				if (productDocCategory.id) {
					let productDocSections = data.sections.filter(
						item => item.category_id === productDocCategory.id
					);

					this.setState({
						items: productDocSections,
						loading: false
					});
				}
			})
			.catch(err => {
				const sidenavFallback = document.getElementById(
					'sidenavFallback'
				);

				this.setState({loading: false});

				sidenavFallback.classList.add('show');
			});
	}

	render({locale}, {items, loading}) {
		if (loading) {
			return <LoadingIndicator />;
		} else {
			return (
				<ul class="nav nav-nested">
					{items.map(item => (
						<li class="nav-item" key={item.id}>
							<ArticlesList
								id={item.id}
								locale={locale}
								name={item.name}
							/>
						</li>
					))}
				</ul>
			);
		}
	}
}