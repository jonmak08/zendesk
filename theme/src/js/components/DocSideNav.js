import preact from 'preact';
import PropTypes from 'prop-types';

import getCN from 'classnames';

import * as apiHelpers from '../helpers/api-helpers';
import LoadingIndicator from './LoadingIndicator';

class ArticlesList extends preact.Component {
	constructor(props) {
		super(props);

		this.getArticles = this.getArticles.bind(this);

		this.state = {
			items: [],
			loading: false
		};
	}

	componentDidMount() {
		if (this.props.expanded) {
			this.getArticles();
		}
	}

	componentDidUpdate(prevProp) {
		const {expanded} = this.props;

		if (prevProp.expanded != expanded && expanded) {
			this.getArticles();
		}
	}

	getArticles() {
		const {id, locale} = this.props;

		this.setState(
			{
				loading: true
			}
		);

		apiHelpers.getArticlesBySectionId(id, locale)
			.then(
				({data}) => {
					this.setState(
						{
							items: data.articles,
							loading: false
						}
					);
				}
			)
			.catch(
				(err) => {
					this.setState(
						{
							loading: false
						}
					);

					if (process.env.NODE_ENV === 'development') {
						console.log(err);
					}
				}
			);
	}

	render({currentArticleId, expanded}, {items, loading}) {
		return (
			<div>
				{expanded && !loading && (
					<ul class="nav nav-nested" role="menu">
						{items.map(
							item => {
								const className = getCN(
									{
										'active': item.id === parseInt(currentArticleId)
									},
									'nav-item'
								);

								return (
									<li class={className} key={item.id} id={item.id} role="none">
										<a class="sidenav-item" href={item.html_url} role="menuitem">
											{item.name}
										</a>
									</li>
								)
							}
						)}
					</ul>
				)}

				{loading && <LoadingIndicator />}
			</div>
		);
	}
}

ArticlesList.PropTypes = {
	currentArticleId: PropTypes.string.isRequired,
	expanded: PropTypes.bool.isRequired,
	id: PropTypes.string.isRequired,
	locale: PropTypes.string.isRequired
};

class DocSideNav extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);

		this.state = {
			expandedItemId: parseInt(this.props.sectionId),
			items: [],
			loading: true
		};
	}

	componentDidMount() {
		const {sectionId, locale} = this.props;

		apiHelpers
			.getSectionBySectionId(sectionId, locale)
			.then(
				({data}) => {
					return (
						apiHelpers.getSectionsByCategoryId(
							data.section.category_id,
							locale
						)
					);
				}
			)
			.then(
				({data}) => {
					this.setState(
						{
							items: data.sections,
							loading: false
						}
					);
				}
			)
			.catch(
				(err) => {
					const sidenavFallback = document.getElementById(
						'sidenavFallback'
					);

					this.setState(
						{
							loading: false
						}
					);

					sidenavFallback.classList.add('show');

					if (process.env.NODE_ENV === 'development') {
						console.log(err);
					}
				}
			);
	}

	handleClick(event) {
		this.setState(
			{
				expandedItemId: parseInt(event.target.id)
			}
		)
	}

	render({currentArticleId, locale}, {expandedItemId, items, loading}) {
		return loading ? (
			<LoadingIndicator />
		) : (
			<ul class="nav nav-nested">
				{items.map(
					item => {
						const expanded = expandedItemId === item.id;

						const className = getCN(
							{
								'expanded': expanded
							},
							'btn-unstyled',
							'sidenav-item'
						);

						return (
							<li class="nav-item" key={item.id}>
								<button aria-haspopup="true" class={className} id={item.id} onClick={this.handleClick} type="button">
									{item.name}
								</button>

								<ArticlesList
									currentArticleId={currentArticleId}
									expanded={expanded}
									id={item.id}
									locale={locale}
								/>
							</li>
						)
					}
				)}
			</ul>
		);
	}
}

DocSideNav.PropTypes = {
	currentArticleId: PropTypes.string.isRequired,
	locale: PropTypes.string.isRequired,
	sectionId: PropTypes.string.isRequired
};

export default DocSideNav;