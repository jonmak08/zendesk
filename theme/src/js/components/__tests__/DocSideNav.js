import preact from 'preact';
import {cleanup, render} from 'preact-testing-library';

import DocSideNav from '../DocSideNav';

afterEach(cleanup);

describe('DocSideNav', () => {
	it('renders correctly', () => {
		const {container} = render(
			<DocSideNav currentArticleId="123" locale="en-us" sectionId="789" />
		);

		expect(container).toMatchSnapshot();
	});
});