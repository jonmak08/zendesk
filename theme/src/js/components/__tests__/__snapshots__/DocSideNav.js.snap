// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocSideNav renders correctly 1`] = `
<div>
  <span
    aria-hidden="true"
    class="loading-animation loading-animation-sm"
  />
</div>
`;
